inst/include/dust/containers.hpp

```diff
@@ -0,0 +1,149 @@
+#ifndef DUST_CONTAINERS_HPP
+#define DUST_CONTAINERS_HPP
+
+#include <cstdint>
+#include <cstddef>
+#include <cstdlib> // malloc
+#include <cstring> // memcpy
+#include <stdexcept>
+#include <sstream>
+#include <vector>
+
+namespace dust {
+
+template <typename T>
+class device_array {
+public:
+  // Default constructor
+  device_array() : data_(nullptr), size_(0) {
+  }
+
+  // Constructor to allocate empty memory
+  device_array(const size_t size) : size_(size) {
+    data_ = (T*) std::malloc(size_ * sizeof(T));
+    if (!data_) {
+      // This is not tested (or easily testable without mocking) but
+      // simple enough. This error will be caught by cpp11
+      //
+      // TODO: we might use `new` here which will throw automatically?
+      throw std::bad_alloc(); // # nocov
```

Comment:
I'm happy if you want to just change these to `new` and `delete`  now – the use of malloc was for direct CUDA conversion, but we know what to do with that so it's less important than it was in the first attempt

Reply:
I'll do this now I think

Reply:
Done in 4605ef6